File: backend/app/api/routes/reports.py
import io
import os
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from openpyxl import Workbook
from reportlab.lib import colors
from reportlab.lib.pagesizes import A4, landscape
from reportlab.lib.units import mm
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from sqlalchemy import func
from sqlalchemy.orm import Session, joinedload

from app.api.deps import get_current_user, scope_disciple_query
from app.core.database import get_db
from app.core.enums import InitiationStatus
from app.models import Disciple, User
from app.schemas.report import CountByKey, ReportSummary

router = APIRouter(prefix="/reports", tags=["reports"])

STATUS_LABELS = {
    InitiationStatus.aspirant: "Аспирант",
    InitiationStatus.recommended: "Рекомендован",
    InitiationStatus.harinama: "Харинама",
    InitiationStatus.brahman: "Брахман",
}

# Register a Unicode (Cyrillic-capable) font for PDF if one is available.
_FONT_CANDIDATES = [
    "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    "/usr/share/fonts/dejavu/DejaVuSans.ttf",
    "/Library/Fonts/Arial Unicode.ttf",
    "/System/Library/Fonts/Supplemental/Arial Unicode.ttf",
    "/System/Library/Fonts/Supplemental/Times New Roman.ttf",
]
PDF_FONT = "Helvetica"
for _path in _FONT_CANDIDATES:
    if os.path.exists(_path):
        try:
            pdfmetrics.registerFont(TTFont("AppFont", _path))
            PDF_FONT = "AppFont"
            break
        except Exception:
            continue


def _apply_filters(q, status_, country, city, temple_id, mentor_id, ready, search):
    if status_:
        q = q.filter(Disciple.initiation_status == status_)
    if country:
        q = q.filter(Disciple.country.ilike(country))
    if city:
        q = q.filter(Disciple.city.ilike(city))
    if temple_id:
        q = q.filter(Disciple.temple_id == temple_id)
    if mentor_id:
        q = q.filter(Disciple.mentor_id == mentor_id)
    if ready is not None:
        q = q.filter(Disciple.ready_for_initiation.is_(ready))
    if search:
        like = f"%{search.strip()}%"
        q = q.filter(Disciple.material_name.ilike(like) | Disciple.spiritual_name.ilike(like))
    return q


def _filtered_query(db, user, status_, country, city, temple_id, mentor_id, ready, search):
    q = scope_disciple_query(
        db.query(Disciple).options(joinedload(Disciple.temple), joinedload(Disciple.mentor)), user
    )
    return _apply_filters(q, status_, country, city, temple_id, mentor_id, ready, search)


# group_by dimension -> (column, label resolver)
def _group_dimension(db, key: str):
    from app.models import Temple

    if key == "status":
        return Disciple.initiation_status, lambda v: STATUS_LABELS.get(v, str(v) if v else "—")
    if key == "country":
        return Disciple.country, lambda v: v or "—"
    if key == "city":
        return Disciple.city, lambda v: v or "—"
    if key == "mentor":
        names = {u.id: u.full_name for u in db.query(User).all()}
        return Disciple.mentor_id, lambda v: names.get(v, "—") if v else "—"
    # default: temple
    tnames = {t.id: t.name for t in db.query(Temple).all()}
    return Disciple.temple_id, lambda v: tnames.get(v, "—") if v else "—"


@router.get("/summary", response_model=ReportSummary)
def summary(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    from app.models import Temple

    total = scope_disciple_query(db.query(Disciple), user).count()

    by_status_rows = (
        scope_disciple_query(db.query(Disciple.initiation_status, func.count()), user)
        .group_by(Disciple.initiation_status)
        .all()
    )
    by_status = [CountByKey(key=STATUS_LABELS.get(s, str(s)), count=c) for s, c in by_status_rows]

    by_country_rows = (
        scope_disciple_query(db.query(Disciple.country, func.count()), user)
        .group_by(Disciple.country)
        .order_by(func.count().desc())
        .all()
    )
    by_country = [CountByKey(key=(c or "—"), count=n) for c, n in by_country_rows]

    by_temple_rows = (
        scope_disciple_query(db.query(Disciple.temple_id, func.count()), user)
        .group_by(Disciple.temple_id)
        .all()
    )
    temple_names = {t.id: t.name for t in db.query(Temple).all()}
    by_temple = [CountByKey(key=(temple_names.get(tid, "—") if tid else "—"), count=n) for tid, n in by_temple_rows]

    ready = scope_disciple_query(db.query(Disciple), user).filter(Disciple.ready_for_initiation.is_(True)).count()

    return ReportSummary(
        total=total, by_status=by_status, by_country=by_country, by_temple=by_temple, ready_for_initiation=ready
    )


@router.get("/group", response_model=list[CountByKey])
def group(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
    group_by: str = Query("status", description="status|country|city|temple|mentor"),
    status_: InitiationStatus | None = Query(None, alias="status"),
    country: str | None = None,
    city: str | None = None,
    temple_id: int | None = None,
    mentor_id: int | None = None,
    ready: bool | None = None,
    q: str | None = None,
):
    col, label = _group_dimension(db, group_by)
    query = scope_disciple_query(db.query(col, func.count()), user)
    query = _apply_filters(query, status_, country, city, temple_id, mentor_id, ready, q)
    rows = query.group_by(col).order_by(func.count().desc()).all()
    return [CountByKey(key=label(v), count=n) for v, n in rows]


_COLUMNS = [
    ("Духовное имя", lambda d: d.spiritual_name or ""),
    ("Мирское имя", lambda d: d.material_name or ""),
    ("Статус", lambda d: STATUS_LABELS.get(d.initiation_status, "")),
    ("Страна", lambda d: d.country or ""),
    ("Город", lambda d: d.city or ""),
    ("Храм", lambda d: d.temple.name if d.temple else ""),
    ("Наставник", lambda d: d.mentor.full_name if d.mentor else ""),
    ("Телефон", lambda d: d.phone or ""),
    ("Email", lambda d: d.email or ""),
]


@router.get("/disciples.xlsx")
def export_xlsx(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
    status_: InitiationStatus | None = Query(None, alias="status"),
    country: str | None = None,
    city: str | None = None,
    temple_id: int | None = None,
    mentor_id: int | None = None,
    ready: bool | None = None,
    q: str | None = None,
):
    rows = (
        _filtered_query(db, user, status_, country, city, temple_id, mentor_id, ready, q)
        .order_by(Disciple.material_name)
        .all()
    )

    wb = Workbook()
    ws = wb.active
    ws.title = "Ученики"
    ws.append([c[0] for c in _COLUMNS])
    for d in rows:
        ws.append([fn(d) for _, fn in _COLUMNS])
    for i, _ in enumerate(_COLUMNS, start=1):
        ws.column_dimensions[ws.cell(row=1, column=i).column_letter].width = 22

    buf = io.BytesIO()
    wb.save(buf)
    buf.seek(0)
    fname = f"disciples_{datetime.now():%Y%m%d}.xlsx"
    return StreamingResponse(
        buf,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f'attachment; filename="{fname}"'},
    )


@router.get("/disciples.pdf")
def export_pdf(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
    status_: InitiationStatus | None = Query(None, alias="status"),
    country: str | None = None,
    city: str | None = None,
    temple_id: int | None = None,
    mentor_id: int | None = None,
    ready: bool | None = None,
    q: str | None = None,
):
    rows = (
        _filtered_query(db, user, status_, country, city, temple_id, mentor_id, ready, q)
        .order_by(Disciple.material_name)
        .all()
    )

    buf = io.BytesIO()
    doc = SimpleDocTemplate(buf, pagesize=landscape(A4), leftMargin=12 * mm, rightMargin=12 * mm,
                            topMargin=12 * mm, bottomMargin=12 * mm)
    styles = getSampleStyleSheet()
    title_style = ParagraphStyle("t", parent=styles["Title"], fontName=PDF_FONT, fontSize=16)
    cell_style = ParagraphStyle("c", parent=styles["Normal"], fontName=PDF_FONT, fontSize=8, leading=10)

    elements = [
        Paragraph("Список учеников — Manibandha", title_style),
        Paragraph(f"Сформировано: {datetime.now():%d.%m.%Y %H:%M} · всего: {len(rows)}", cell_style),
        Spacer(1, 6 * mm),
    ]

    header = [Paragraph(f"<b>{c[0]}</b>", cell_style) for c in _COLUMNS]
    data = [header]
    for d in rows:
        data.append([Paragraph(str(fn(d)), cell_style) for _, fn in _COLUMNS])

    table = Table(data, repeatRows=1)
    table.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#7c2d12")),
                ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
                ("GRID", (0, 0), (-1, -1), 0.25, colors.HexColor("#d6c3a5")),
                ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, colors.HexColor("#faf6f0")]),
                ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
                ("FONTNAME", (0, 0), (-1, -1), PDF_FONT),
            ]
        )
    )
    elements.append(table)
    doc.build(elements)
    buf.seek(0)

    fname = f"disciples_{datetime.now():%Y%m%d}.pdf"
    return StreamingResponse(
        buf, media_type="application/pdf", headers={"Content-Disposition": f'attachment; filename="{fname}"'}
    )

File: backend/internal/web/reports.go
package web

import (
	"bytes"
	_ "embed"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/go-pdf/fpdf"
	"github.com/xuri/excelize/v2"
	"gorm.io/gorm"

	"manibandha/internal/models"
)

//go:embed fonts/DejaVuSans.ttf
var dejavuTTF []byte

var reportStatusLabels = map[string]string{
	"aspirant":    "Кандидат",
	"pranama":     "Пранама-мантра",
	"recommended": "Зарегистрирован",
	"harinama":    "Харинама",
	"brahman":     "Брахман",
}

func statusLabel(s string) string {
	if l, ok := reportStatusLabels[s]; ok {
		return l
	}
	return s
}

// applyReportFilters — те же фильтры, что в reports.py (без status при group по status передаётся отдельно).
func applyReportFilters(q *gorm.DB, qp map[string]string) *gorm.DB {
	if v := qp["status"]; v != "" {
		q = q.Where("initiation_status = ?", v)
	}
	if v := qp["country"]; v != "" {
		q = q.Where("country ILIKE ?", v)
	}
	if v := qp["region"]; v != "" {
		q = q.Where("region ILIKE ?", v)
	}
	if v := qp["city"]; v != "" {
		q = q.Where("city ILIKE ?", v)
	}
	if v := qp["temple_id"]; v != "" {
		q = q.Where("temple_id = ?", v)
	}
	if v := qp["mentor_id"]; v != "" {
		q = q.Where("mentor_id = ?", v)
	}
	if v := qp["ready"]; v != "" {
		q = q.Where("ready_for_initiation = ?", v == "true")
	}
	if v := strings.TrimSpace(qp["q"]); v != "" {
		like := "%" + v + "%"
		q = q.Where("material_name ILIKE ? OR spiritual_name ILIKE ?", like, like)
	}
	return q
}

// GET /reports/summary
func (s *Server) reportSummary(w http.ResponseWriter, r *http.Request) {
	u := currentUser(r)
	base := func() *gorm.DB { return scopeDisciples(s.db(r).Model(&models.Disciple{}), u) }

	var total int64
	base().Count(&total)

	type sc struct {
		K string
		C int64
	}
	var statusRows []sc
	base().Select("initiation_status as k, count(*) as c").Group("initiation_status").Scan(&statusRows)
	byStatus := make([]map[string]any, 0, len(statusRows))
	for _, x := range statusRows {
		byStatus = append(byStatus, map[string]any{"key": statusLabel(x.K), "count": x.C})
	}

	type cc struct {
		K *string
		C int64
	}
	var countryRows []cc
	base().Select("country as k, count(*) as c").Group("country").Order("count(*) DESC").Scan(&countryRows)
	byCountry := make([]map[string]any, 0, len(countryRows))
	for _, x := range countryRows {
		byCountry = append(byCountry, map[string]any{"key": strOrDash(x.K), "count": x.C})
	}

	type tc struct {
		K *int
		C int64
	}
	var templeRows []tc
	base().Select("temple_id as k, count(*) as c").Group("temple_id").Scan(&templeRows)
	tnames := s.templeNames(r)
	byTemple := make([]map[string]any, 0, len(templeRows))
	for _, x := range templeRows {
		name := "—"
		if x.K != nil {
			if n, ok := tnames[*x.K]; ok {
				name = n
			}
		}
		byTemple = append(byTemple, map[string]any{"key": name, "count": x.C})
	}

	var ready, readyPr int64
	base().Where("ready_for_initiation = ?", true).Count(&ready)
	base().Where("ready_for_pranama = ?", true).Count(&readyPr)

	writeJSON(w, http.StatusOK, map[string]any{
		"total": total, "by_status": byStatus, "by_country": byCountry, "by_temple": byTemple,
		"ready_for_pranama": readyPr, "ready_for_initiation": ready,
	})
}

func strOrDash(p *string) string {
	if p == nil || *p == "" {
		return "—"
	}
	return *p
}

func (s *Server) templeNames(r *http.Request) map[int]string {
	var temples []models.Temple
	s.db(r).Find(&temples)
	m := map[int]string{}
	for _, t := range temples {
		m[t.ID] = t.Name
	}
	return m
}

func (s *Server) discipleNames(r *http.Request) map[int]string {
	var ds []models.Disciple
	s.db(r).Select("id, spiritual_name, material_name").Find(&ds)
	m := map[int]string{}
	for i := range ds {
		m[ds[i].ID] = ds[i].Name()
	}
	return m
}

// GET /reports/timeline
func (s *Server) reportTimeline(w http.ResponseWriter, r *http.Request) {
	u := currentUser(r)
	type dts struct {
		PranamaDate  *time.Time
		HarinamaDate *time.Time
		BrahmanDate  *time.Time
	}
	var rows []dts
	scopeDisciples(s.db(r).Model(&models.Disciple{}), u).
		Select("pranama_date, harinama_date, brahman_date").Scan(&rows)

	buckets := map[string]map[string]int{}
	add := func(t *time.Time, kind string) {
		if t == nil {
			return
		}
		key := t.Format("2006-01")
		if buckets[key] == nil {
			buckets[key] = map[string]int{"pranama": 0, "harinama": 0, "brahman": 0}
		}
		buckets[key][kind]++
	}
	for _, x := range rows {
		add(x.PranamaDate, "pranama")
		add(x.HarinamaDate, "harinama")
		add(x.BrahmanDate, "brahman")
	}
	keys := make([]string, 0, len(buckets))
	for k := range buckets {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	out := make([]map[string]any, 0, len(keys))
	for _, k := range keys {
		b := buckets[k]
		out = append(out, map[string]any{"period": k, "pranama": b["pranama"], "harinama": b["harinama"], "brahman": b["brahman"]})
	}
	writeJSON(w, http.StatusOK, out)
}

// GET /reports/group
func (s *Server) reportGroup(w http.ResponseWriter, r *http.Request) {
	u := currentUser(r)
	qp := queryMap(r)
	groupBy := qp["group_by"]
	if groupBy == "" {
		groupBy = "status"
	}
	col := map[string]string{
		"status": "initiation_status", "country": "country", "region": "region",
		"city": "city", "mentor": "mentor_id", "temple": "temple_id",
	}[groupBy]
	if col == "" {
		col = "temple_id"
	}

	q := scopeDisciples(s.db(r).Model(&models.Disciple{}), u)
	q = applyReportFilters(q, qp)
	type row struct {
		V *string
		C int64
	}
	var rows []row
	q.Select(col + " as v, count(*) as c").Group(col).Order("count(*) DESC").Scan(&rows)

	var label func(*string) string
	switch groupBy {
	case "status":
		label = func(v *string) string {
			if v == nil {
				return "—"
			}
			return statusLabel(*v)
		}
	case "country", "region", "city":
		label = func(v *string) string { return strOrDash(v) }
	case "mentor":
		names := s.discipleNames(r)
		label = func(v *string) string { return lookupName(v, names) }
	default: // temple
		names := s.templeNames(r)
		label = func(v *string) string { return lookupName(v, names) }
	}
	out := make([]map[string]any, 0, len(rows))
	for _, x := range rows {
		out = append(out, map[string]any{"key": label(x.V), "count": x.C})
	}
	writeJSON(w, http.StatusOK, out)
}

func lookupName(v *string, names map[int]string) string {
	if v == nil || *v == "" {
		return "—"
	}
	var id int
	fmt.Sscanf(*v, "%d", &id)
	if n, ok := names[id]; ok {
		return n
	}
	return "—"
}

func queryMap(r *http.Request) map[string]string {
	m := map[string]string{}
	for k, v := range r.URL.Query() {
		if len(v) > 0 {
			m[k] = v[0]
		}
	}
	return m
}

// ── экспорт ─────────────────────────────────────────────────────────────────

var reportColumns = []struct {
	Head string
	Val  func(d *models.Disciple) string
}{
	{"Духовное имя", func(d *models.Disciple) string { return deref(d.SpiritualName) }},
	{"Мирское имя", func(d *models.Disciple) string { return d.MaterialName }},
	{"Статус", func(d *models.Disciple) string { return statusLabel(d.InitiationStatus) }},
	{"Страна", func(d *models.Disciple) string { return deref(d.Country) }},
	{"Область", func(d *models.Disciple) string { return deref(d.Region) }},
	{"Город", func(d *models.Disciple) string { return deref(d.City) }},
	{"Храм", func(d *models.Disciple) string {
		if d.Temple != nil {
			return d.Temple.Name
		}
		return ""
	}},
	{"Куратор", func(d *models.Disciple) string {
		if d.Mentor != nil {
			return d.Mentor.Name()
		}
		return ""
	}},
	{"Телефон", func(d *models.Disciple) string { return deref(d.Phone) }},
	{"Email", func(d *models.Disciple) string { return deref(d.Email) }},
}

func deref(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

func (s *Server) reportRows(r *http.Request, u *models.User, qp map[string]string) []models.Disciple {
	q := scopeDisciples(s.db(r).Preload("Temple").Preload("Mentor"), u)
	q = applyReportFilters(q, qp)
	var rows []models.Disciple
	q.Order("material_name").Find(&rows)
	return rows
}

// GET /reports/disciples.xlsx
func (s *Server) exportXlsx(w http.ResponseWriter, r *http.Request) {
	rows := s.reportRows(r, currentUser(r), queryMap(r))
	f := excelize.NewFile()
	sheet := "Ученики"
	f.SetSheetName(f.GetSheetName(0), sheet)
	for i, c := range reportColumns {
		cell, _ := excelize.CoordinatesToCellName(i+1, 1)
		f.SetCellStr(sheet, cell, c.Head)
		col, _ := excelize.ColumnNumberToName(i + 1)
		f.SetColWidth(sheet, col, col, 22)
	}
	for ri := range rows {
		for ci, c := range reportColumns {
			v := c.Val(&rows[ri])
			if v == "" {
				continue // пустые ячейки не пишем — как openpyxl (None)
			}
			cell, _ := excelize.CoordinatesToCellName(ci+1, ri+2)
			f.SetCellStr(sheet, cell, v)
		}
	}
	var buf bytes.Buffer
	f.Write(&buf)
	fname := fmt.Sprintf("disciples_%s.xlsx", time.Now().Format("20060102"))
	w.Header().Set("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, fname))
	w.Write(buf.Bytes())
}

// GET /reports/disciples.pdf
func (s *Server) exportPdf(w http.ResponseWriter, r *http.Request) {
	rows := s.reportRows(r, currentUser(r), queryMap(r))
	pdf := fpdf.New("L", "mm", "A4", "")
	pdf.SetMargins(12, 12, 12)
	pdf.AddPage()

	// Встроенный DejaVuSans (кириллица на любом хосте, без зависимости от системных шрифтов).
	font := "Helvetica"
	if len(dejavuTTF) > 0 {
		pdf.AddUTF8FontFromBytes("app", "", dejavuTTF)
		font = "app"
	}
	pdf.SetFont(font, "", 16)
	pdf.CellFormat(0, 8, "Список учеников — Е.М. Манибандха Прабху", "", 1, "L", false, 0, "")
	pdf.SetFont(font, "", 8)
	pdf.CellFormat(0, 5, fmt.Sprintf("Сформировано: %s · всего: %d", time.Now().Format("02.01.2006 15:04"), len(rows)), "", 1, "L", false, 0, "")
	pdf.Ln(4)

	// ширины колонок (сумма ≈ 273 мм полезной ширины A4 landscape)
	widths := []float64{34, 34, 26, 24, 26, 24, 30, 30, 26, 30}
	pdf.SetFillColor(200, 116, 42)
	pdf.SetTextColor(255, 255, 255)
	for i, c := range reportColumns {
		pdf.CellFormat(widths[i], 7, c.Head, "1", 0, "L", true, 0, "")
	}
	pdf.Ln(-1)
	pdf.SetTextColor(0, 0, 0)
	fill := false
	for ri := range rows {
		if fill {
			pdf.SetFillColor(250, 246, 240)
		} else {
			pdf.SetFillColor(255, 255, 255)
		}
		for i, c := range reportColumns {
			pdf.CellFormat(widths[i], 6, c.Val(&rows[ri]), "1", 0, "L", true, 0, "")
		}
		pdf.Ln(-1)
		fill = !fill
	}

	var buf bytes.Buffer
	pdf.Output(&buf)
	fname := fmt.Sprintf("disciples_%s.pdf", time.Now().Format("20060102"))
	w.Header().Set("Content-Type", "application/pdf")
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, fname))
	w.Write(buf.Bytes())
}
